### PR TITLE
feat: additional request and response rules (DPE-106)

### PR DESCRIPTION
**Pull request by brandonsahadeo**

### CODE REVIEW

File: rulesets/src/request-response.ruleset.yml [hunk 1]
```diff
@@ -68,7 +76,80 @@ rules:
       functionOptions:
         notMatch: "^(x|X)-"
 
+  sps-invalid-location-header:
+    description: Location header must not be present in non-201 responses
```
Comment (travisgosselin) [inline rulesets/src/request-response.ruleset.yml:80]:
```suggestion
    description: Location header MUST not be present in non-201 responses
```

File: rulesets/src/request-response.ruleset.yml [hunk 2]
```diff
@@ -68,7 +76,80 @@ rules:
...
+    description: Security field MUST be present with at least one item
+    severity: error
+    given: "$"
+    then:
+      field: "security"
+      function: schema
+      functionOptions:
+        schema:
+          type: array
+          minItems: 1
```
Comment (travisgosselin) [inline rulesets/src/request-response.ruleset.yml:97]: does this enforce that the security field is defined? or just if it is defined that it has 1?

File: rulesets/src/request-response.ruleset.yml [hunk 3]
```diff
@@ -68,7 +76,80 @@ rules:
...
+    description: Content-Language is optional but MUST default locale to en-US when none provided
+    severity: error
+    given: $.paths[*][*].responses[*].headers.Content-Language.schema.default
+    then:
+      function: pattern
+      functionOptions:
+        match: "en-US"
+
+  sps-no-explicit-headers:
+    description: Access-Control-*, Content-Type, and Accept Headers should not be specified explicitly in a spec as it an operational concern.
```
Comment (travisgosselin) [inline rulesets/src/request-response.ruleset.yml:109]:
```suggestion
    description: Access-Control-*, Content-Type, and Accept Headers SHOULD NOT be specified explicitly in a spec as it an operational concern.
```

File: rulesets/src/request-response.ruleset.yml [hunk 4]
```diff
@@ -68,7 +76,80 @@ rules:
...
+    description: Custom headers MUST NOT be longer than 50 chars, and MUST only contain alphanumeric and dash chars, and MUST begin with SPS- not x-
+    severity: error
+    given: "$.paths[*][*].responses[*].headers.*~"
+    then:
+      function: pattern
+      functionOptions:
+        match: "^(SPS-[a-zA-Z0-9-]{1,50}|Accept|Access-Control.*|Authorization|Cache-Control|Content-Language|Content-Length|Content-Type|Date|ETag|Host|If-Match|If-None-Match|Location|Origin|User-Agent)$"
+
+  sps-sensitive-data-in-headers:
+    description: Headers MUST NOT contain sensitive data
```
Comment (travisgosselin) [inline rulesets/src/request-response.ruleset.yml:127]:
```suggestion
    description: Headers MUST NOT contain sensitive data.
```

File: rulesets/src/request-response.ruleset.yml [hunk 5]
```diff
@@ -68,7 +76,80 @@ rules:
...
+
   ##### MIME Types #####
+  sps-no-resource-extensions:
+    description: Request and Response media type formats MUST NOT be implied using extensions on resources (i.e. .json or .xml). Instead, use the standard resource path with the appropriate Content-Type header
+    severity: error
+    given: "$.paths.*~"
+    then:
+      function: pattern
+      functionOptions:
+        notMatch: "\\.json|\\.xml"
```
Comment (travisgosselin) [inline rulesets/src/request-response.ruleset.yml:143]: lets include .yml and .yaml here too

**Review by travisgosselin — COMMENTED** [review]
Nice additions.

File: rulesets/src/request-response.ruleset.yml (hunk 2 above)
Comment (brandonsahadeo) [reply rulesets/src/request-response.ruleset.yml:97]: It enforces that its defined with at least 1 item in the array.

File: rulesets/test/request-response/sps-invalid-custom-header-format.test.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                  SPS-Custom-Header-123456789012345678901234567890123456789012345678901:
+                    schema:
+                      type: string
+                      default: custom value
+    `;
+
+    await spectral.validateFailure(spec, ruleName, "Error", 1);
+  });
+
+  test("custom headers cant start with X-", async () => {
```
Comment (markdebeer) [inline rulesets/test/request-response/sps-invalid-custom-header-format.test.js:50]: Just missing the apostrophe in can't ->
```suggestion
  test("custom headers can't start with X-", async () => {
```

**Review by markdebeer — APPROVED** [review]

**Pull request merged**